### PR TITLE
planner: choose TableScan when use an empty index hint

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

fix https://github.com/pingcap/tidb/issues/12033.

schema:
```
mysql> create table t(a int, b int primary key, index idx_a(a));
```

before this pr:
```
mysql> explain select a from t use index();
+-------------------+----------+------+---------------------------------------------------------------------+
| id                | count    | task | operator info                                                       |
+-------------------+----------+------+---------------------------------------------------------------------+
| IndexReader_7     | 10000.00 | root | index:IndexScan_6                                                   |
| └─IndexScan_6     | 10000.00 | cop  | table:t, index:a, range:[NULL,+inf], keep order:false, stats:pseudo |
+-------------------+----------+------+---------------------------------------------------------------------+
```

after this pr:
```
mysql> explain select a from t use index();
+-------------------+----------+------+------------------------------------------------------------+
| id                | count    | task | operator info                                              |
+-------------------+----------+------+------------------------------------------------------------+
| TableReader_5     | 10000.00 | root | data:TableScan_4                                           |
| └─TableScan_4     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+-------------------+----------+------+------------------------------------------------------------+
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -558,6 +558,19 @@ func (b *PlanBuilder) getPossibleAccessPaths(indexHints []*ast.IndexHint, tblInf
 		}
 
 		hasScanHint = true
+		if hint.IndexNames == nil {
+			// If indexNames is nil, choose TablePath.
+			// For example: `select a from t use index()`.
+			if path := getTablePath(publicPaths); path != nil {
+				if hint.HintType == ast.HintIgnore {
+					ignored = append(ignored, path)
```
Comment (winoros) [inline planner/core/planbuilder.go]: This ignore case doesn't make sense?

Comment (francis0407) [reply planner/core/planbuilder.go]: Sure. I'll remove this.

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -558,6 +558,19 @@ func (b *PlanBuilder) getPossibleAccessPaths(indexHints []*ast.IndexHint, tblInf
 		}
 
 		hasScanHint = true
+		if hint.IndexNames == nil {
```
Comment (lamxTyler) [inline planner/core/planbuilder.go]: Could you also consider the comment style hint like `index(t)`?

File: planner/core/planbuilder.go (hunk 1 above)
Comment (francis0407) [reply planner/core/planbuilder.go]: I think this is another issue. I tried `ignore index()` in MySQL, it returns a syntax Error. 
```
mysql> explain select b from tt ignore index();
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ')' at line 1
```
But TiDB allows such syntax.

File: planner/core/planbuilder.go (hunk 2 above)
Comment (francis0407) [reply planner/core/planbuilder.go]: fixed

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -558,6 +558,15 @@ func (b *PlanBuilder) getPossibleAccessPaths(indexHints []*ast.IndexHint, tblInf
...
+			// If indexNames is nil, choose TablePath.
+			// For example: `select a from t use index()`.
+			if path := getTablePath(publicPaths); path != nil {
+				hasUseOrForce = true
+				path.forced = true
```
Comment (winoros) [inline planner/core/planbuilder.go]: You need check whether is ignore or force.
Or you'll set table path as forced although no index is ignored by `IGNORE INDEX()`

Comment (francis0407) [reply planner/core/planbuilder.go]: fixed

File: planner/core/planbuilder.go [hunk 4]
```diff
@@ -558,6 +558,15 @@ func (b *PlanBuilder) getPossibleAccessPaths(indexHints []*ast.IndexHint, tblInf
 		}
 
 		hasScanHint = true
+		if hint.IndexNames == nil && hint.HintType != ast.HintIgnore {
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: How about adding a comment? From [MySQL document](https://dev.mysql.com/doc/refman/8.0/en/index-hints.html):
> It is syntactically valid to omit index_list for USE INDEX, which means “use no indexes.” Omitting index_list for FORCE INDEX or IGNORE INDEX is a syntax error.

Comment (francis0407) [reply planner/core/planbuilder.go]: addressed

**Pull request merged**